### ISSUE TITLE
Sap.Data.Hana.HanaException:  'Parameter/Column (4) not bound.'

### ISSUE THREAD

**Issue by MichielDG** [linq2db/linq2db#1528]
I've using linq2db for SapHana  and when I've updated to linq2db version 2.x there is an issue with an query. When I revert to linq2db version 1.10.2, my query works again.
```
Exception message: 'Parameter/Column (3) not bound.'
Stack trace:       at Sap.Data.Hana.HanaCommand._ExecuteReader(CommandBehavior commandBehavior, Boolean isExecuteScalar, Boolean isBeginExecuteReader)
   at Sap.Data.Hana.HanaCommand.ExecuteReaderBatch(CommandBehavior behavior)
   at Sap.Data.Hana.HanaCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader(CommandBehavior behavior)
   at LinqToDB.Data.DataConnection.ExecuteReader(CommandBehavior commandBehavior) in C:\projects\linq2db\Source\LinqToDB\Data\DataConnection.cs:line 1227
   at LinqToDB.Data.DataConnection.QueryRunner.ExecuteReader() in C:\projects\linq2db\Source\LinqToDB\Data\DataConnection.QueryRunner.cs:line 445
   at LinqToDB.Linq.QueryRunner.<ExecuteQuery>d__9`1.MoveNext() in C:\projects\linq2db\Source\LinqToDB\Linq\QueryRunner.cs:line 320
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at linq2dbHanaTest.Program.Main(String[] args) in C:\Users\user\source\repos\linq2dbHanaTest\linq2dbHanaTest\Program.cs:line 36
```

### Steps to reproduce
Execute the query.

```c#
using (var db = new SboDataConnection(connectionString, ProviderName.SapHana))
{
    var query =
        db.OITM.Where(x =>
                (x.validFor.Equals("Y") && (x.validFrom == null || x.validFrom <= dateTime) && (x.validTo == null || x.validTo >= date)) ||
                (x.frozenFor.Equals("Y") && (x.frozenFrom == null || x.frozenFrom <= dateTime) && (x.frozenTo == null || x.frozenTo >= date)) ||
                (x.frozenFor.Equals("N") && x.validFor.Equals("N"))
            ).Select(x => new
            {
                Code = x.ItemCode,
                Name = x.ItemName
            });
    query.ToList();
}
```

### Environment details
linq2db version: *2.6.2*
Database Server: *Hana 1.00*
Database Provider: *Sap.Data.Hana.v4.5*
Framework version: *.NET Framework 4.7.2*

**Comment by sdanyliv** [comment]
@MichielDG, which SQL was generated? 

I have idea that it is because we have optimized out parameters.
Can you try this query:
```cs
    var dateTime1 = dateTime;
    var dateTime2 = dateTime;
    var date1 = date;
    var date2 = date;

    var query =
        db.OITM.Where(x =>
                (x.validFor.Equals("Y") && (x.validFrom == null || x.validFrom <= dateTime1) && (x.validTo == null || x.validTo >= date1) ||
                (x.frozenFor.Equals("Y") && (x.frozenFrom == null || x.frozenFrom <= dateTime2) && (x.frozenTo == null || x.frozenTo >= date2)) ||
                (x.frozenFor.Equals("N") && x.validFor.Equals("N"))
            ).Select(x => new
            {
                Code = x.ItemCode,
                Name = x.ItemName
            });
    query.ToList();
```

**Comment by MichielDG** [comment]
@sdanyliv 

When I use for each parameter another variable like you suggest I works.

The SQL that is generated:
**With using variables multiple times:**
``` SQL
--  SapHana
DECLARE @dateTime  -- DateTime2
SET     @dateTime = '2019-01-11 15:05:06.117'
DECLARE @date0  -- DateTime2
SET     @date0 = '2019-01-11'

SELECT
	"x"."ItemCode",
	"x"."ItemName"
FROM
	"OITM" "x"
WHERE
	("x"."validFor" = 'Y' AND ("x"."validFrom" IS NULL OR "x"."validFrom" <= :dateTime) AND ("x"."validTo" IS NULL OR "x"."validTo" >= :date0) OR "x"."frozenFor" = 'Y' AND ("x"."frozenFrom" IS NULL OR "x"."frozenFrom" <= :dateTime) AND ("x"."frozenTo" IS NULL OR "x"."frozenTo" >= :date0)) OR
	"x"."frozenFor" = 'N' AND "x"."validFor" = 'N'
```
**SQL of your solution**
``` SQL
--  SapHana
DECLARE @dateTime1  -- DateTime2
SET     @dateTime1 = '2019-01-11 15:08:08.748'
DECLARE @date1  -- DateTime2
SET     @date1 = '2019-01-11'
DECLARE @dateTime2  -- DateTime2
SET     @dateTime2 = '2019-01-11 15:08:08.748'
DECLARE @date2  -- DateTime2
SET     @date2 = '2019-01-11'

SELECT
	"x"."ItemCode",
	"x"."ItemName"
FROM
	"OITM" "x"
WHERE
	("x"."validFor" = 'Y' AND ("x"."validFrom" IS NULL OR "x"."validFrom" <= :dateTime1) AND ("x"."validTo" IS NULL OR "x"."validTo" >= :date1) OR "x"."frozenFor" = 'Y' AND ("x"."frozenFrom" IS NULL OR "x"."frozenFrom" <= :dateTime2) AND ("x"."frozenTo" IS NULL OR "x"."frozenTo" >= :date2)) OR
	"x"."frozenFor" = 'N' AND "x"."validFor" = 'N'
```

Will this be fixed in a future version?

**Comment by sdanyliv** [comment]
@MichielDG, yes. Preparing fix now. It will be in next minor update, I think next Thursday.

**Comment by MaceWindu** [comment]
Fixed by #1530
